File: parsers/go/evaluate_test.go
package kato

import (
	"reflect"
	"testing"
)

func TestEvaluate(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		opts     []EvalOption
		expected map[string]any
	}{
		{
			name:  "basic object with integer",
			input: "server {\n  port: 3000\n}",
			expected: map[string]any{
				"server": map[string]any{
					"port": 3000,
				},
			},
		},
		{
			name:  "nested objects",
			input: "db {\n  pool {\n    min: 2\n  }\n}",
			expected: map[string]any{
				"db": map[string]any{
					"pool": map[string]any{
						"min": 2,
					},
				},
			},
		},
		{
			name:  "arrays",
			input: `regions: ["a", "b"]`,
			expected: map[string]any{
				"regions": []any{"a", "b"},
			},
		},
		{
			name:  "string scalar",
			input: `name: "hello"`,
			expected: map[string]any{
				"name": "hello",
			},
		},
		{
			name:  "integer scalar",
			input: "count: 42",
			expected: map[string]any{
				"count": 42,
			},
		},
		{
			name:  "float scalar",
			input: "rate: 3.14",
			expected: map[string]any{
				"rate": 3.14,
			},
		},
		{
			name:  "bool true",
			input: "enabled: true",
			expected: map[string]any{
				"enabled": true,
			},
		},
		{
			name:  "bool false",
			input: "enabled: false",
			expected: map[string]any{
				"enabled": false,
			},
		},
		{
			name:  "null",
			input: "value: null",
			expected: map[string]any{
				"value": nil,
			},
		},
		{
			name:  "on becomes true",
			input: "enabled: on",
			expected: map[string]any{
				"enabled": true,
			},
		},
		{
			name:  "off becomes false",
			input: "debug: off",
			expected: map[string]any{
				"debug": false,
			},
		},
		{
			name:  "bare word",
			input: "mode: production",
			expected: map[string]any{
				"mode": "production",
			},
		},
		{
			name:  "unit literal",
			input: "timeout: 30s",
			expected: map[string]any{
				"timeout": map[string]any{
					"$unit": "s",
					"value": 30.0,
				},
			},
		},
		{
			name:  "env with option",
			input: `host: env("DB_HOST")`,
			opts:  []EvalOption{WithEnv(map[string]string{"DB_HOST": "localhost"})},
			expected: map[string]any{
				"host": "localhost",
			},
		},
		{
			name:  "ref resolves internal reference",
			input: "base_port: 3000\nport: ref(\"base_port\")",
			expected: map[string]any{
				"base_port": 3000,
				"port":      3000,
			},
		},
		{
			name:  "profile merges overrides",
			input: "server {\n  port: 3000\n}\n@profile production {\n  server {\n    port: 8080\n  }\n}",
			opts:  []EvalOption{WithProfile("production")},
			expected: map[string]any{
				"server": map[string]any{
					"port": 8080,
				},
			},
		},
		{
			name:  "top-level keys without wrapping object",
			input: "host: \"localhost\"\nport: 5432",
			expected: map[string]any{
				"host": "localhost",
				"port": 5432,
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := Evaluate(tt.input, tt.opts...)
			if err != nil {
				// Stub returns "not implemented" — mark as expected for now
				t.Skipf("Evaluate not yet implemented: %v", err)
			}
			if !reflect.DeepEqual(result, tt.expected) {
				t.Errorf("Evaluate() = %v, want %v", result, tt.expected)
			}
		})
	}
}

File: parsers/ts/tests/evaluate.test.ts
import { describe, it, expect } from "vitest";
import { evaluate } from "../src/evaluate.js";

describe("evaluate()", () => {
  describe("basic objects", () => {
    it("evaluates a simple named block", () => {
      expect(evaluate(`server { port: 3000 }`)).toEqual({ server: { port: 3000 } });
    });

    it("evaluates nested blocks", () => {
      expect(evaluate(`db { pool { min: 2 } }`)).toEqual({ db: { pool: { min: 2 } } });
    });
  });

  describe("arrays", () => {
    it("evaluates inline arrays", () => {
      expect(evaluate(`regions: ["a", "b"]`)).toEqual({ regions: ["a", "b"] });
    });
  });

  describe("scalars", () => {
    it("evaluates strings", () => {
      expect(evaluate(`name: "hello"`)).toEqual({ name: "hello" });
    });

    it("evaluates integers", () => {
      expect(evaluate(`count: 42`)).toEqual({ count: 42 });
    });

    it("evaluates floats", () => {
      expect(evaluate(`ratio: 3.14`)).toEqual({ ratio: 3.14 });
    });

    it("evaluates booleans", () => {
      expect(evaluate(`active: true`)).toEqual({ active: true });
    });

    it("evaluates null", () => {
      expect(evaluate(`value: null`)).toEqual({ value: null });
    });

    it("evaluates on as true", () => {
      expect(evaluate(`enabled: on`)).toEqual({ enabled: true });
    });

    it("evaluates off as false", () => {
      expect(evaluate(`disabled: off`)).toEqual({ disabled: false });
    });

    it("evaluates bare words", () => {
      expect(evaluate(`mode: production`)).toEqual({ mode: "production" });
    });
  });

  describe("unit literals", () => {
    it("evaluates unit literals to $unit object", () => {
      expect(evaluate(`timeout: 30s`)).toEqual({ timeout: { $unit: "s", value: 30 } });
    });
  });

  describe("env()", () => {
    it("resolves env() calls with provided env option", () => {
      const result = evaluate(`url: env("DB_URL")`, { env: { DB_URL: "postgres://localhost/db" } });
      expect(result).toEqual({ url: "postgres://localhost/db" });
    });
  });

  describe("ref()", () => {
    it("resolves ref() to the referenced value", () => {
      const input = `
shared { region: "us-east-1" }
deploy { region: ref("shared.region") }
`;
      expect(evaluate(input)).toEqual({
        shared: { region: "us-east-1" },
        deploy: { region: "us-east-1" },
      });
    });
  });

  describe("profiles", () => {
    it("merges profile overrides when profile option is set", () => {
      const input = `
server { port: 3000 }
#[profile "production"]
server { port: 8080 }
`;
      const result = evaluate(input, { profile: "production" });
      expect(result).toEqual({ server: { port: 8080 } });
    });
  });

  describe("top-level keys", () => {
    it("handles multiple top-level key-value pairs", () => {
      const input = `
host: "localhost"
port: 5432
`;
      expect(evaluate(input)).toEqual({ host: "localhost", port: 5432 });
    });
  });
});
